### ISSUE TITLE
Beta.9 new namespace breaking css parent selector

### ISSUE THREAD

**Issue by marmite22** [styled-components/styled-components#3909]
### System:
 - OS: macOS 12.5.1
 - CPU: (10) arm64 Apple M1 Pro
 - Memory: 78.30 MB / 32.00 GB
 - Shell: 5.8.1 - /bin/zsh
### Binaries:
 - Node: 18.7.0 - /opt/homebrew/bin/node
 - Yarn: 1.22.19 - /opt/homebrew/bin/yarn
 - npm: 8.15.0 - /opt/homebrew/bin/npm

## Reproduction

Using Beta 9. 
```
import React from "react";
import styled, { StyleSheetManager } from "styled-components";

const StyledComponent = styled.div`
  color: blue;
  .child & {
    color: green;
  }
`;

const App = () => {
  return (
    <div className="parent">
      <div className="child">
        <StyledComponent>Green</StyledComponent>
      </div>
      <StyledComponent>Blue</StyledComponent>
    </div>
  );
};

export default () => {
  return (
    <StyleSheetManager namespace={".parent"}>
      <App />
    </StyleSheetManager>
  );
};

```

## Steps to reproduce
https://codesandbox.io/s/parent-ampersand-u9hgop?file=/src/App.js
I have created a PR and added a failing test but wasn't able to solve the issue myself: https://github.com/styled-components/styled-components/pull/3908

## Expected Behavior
I would expect the above to generate 
```
.parent .child .sc-a{color:green;}
```

## Actual Behavior
but instead I get 
```
.child .parent .sc-a{color:green;}
```
Note the child selector is moving right to the front, before the namespace.

**Comment by probablyup** [comment]
Yup this is a bug. The relevant code is in here: https://github.com/styled-components/styled-components/blob/d21a7af6435336f10b7e8d891eef253e1ad30051/packages/styled-components/src/utils/stylis.ts

**Comment by marmite22** [comment]
Thanks, I did raise a PR https://github.com/styled-components/styled-components/pull/3908 which you have responded to but I think it's got a bit confused because I uncovered a second bug while I was fixing it.

In summary this issue is fixed by the PR but I added 2 tests and the 2nd test fails because of an unrelated bug.

**Comment by vadymshymko** [comment]
Hi there! Is there any updates?

I faced similar problem:

When using the ampersand and the element>element selector (`& > someTag`) in styles that are based on a specific value of some prop (e.g. `$test`), the styles are applied not specifically to that one instance of the styled component, but to all, even to those instances whose $test value does not match the condition for applying styles.

Example at the [link](https://codesandbox.io/s/styled-components-v6-beta-sandbox-forked-mpc9h7?file=/src/App.tsx).
Expected behavior: when clicking on the first red square, the text color of 'Foo Title' becomes black, the text color of 'Bar Title' remains white

Actual behavior: when clicking on the first red square, the text color of 'Foo Title' turns black, the text color of 'Bar Title' also turns black

**Comment by marmite22** [comment]
@vadymshymko hey, I submitted a PR that I hope fixes the issue https://github.com/styled-components/styled-components/pull/3908 - could you take a look at the tests I included in that PR and see if they are an accurate test of the issue you are seeing.

**Comment by vadymshymko** [comment]
> @vadymshymko hey, I submitted a PR that I hope fixes the issue #3908 - could you take a look at the tests I included in that PR and see if they are an accurate test of the issue you are seeing.

@marmite22  thank you. I can’t be 100% sure yet, but I hope this PR will help solve the problem. I'll wait for the release and check)

**Comment by marmite22** [comment]
@probablyup - now that the fix for this is merged in is there any chance you can release another beta so I can test it out at work (where I originally spotted the issue)?

**Comment by probablyup** [comment]
@marmite22 yes, I'm hoping to resolve a few type errors in my PR before I cut another beta

**Comment by quantizor** [comment]
Closing — this issue is stale (2+ years without resolution) and relates to an older version of styled-components. If this is still reproducible on the latest v6.x release, please open a new issue with a minimal reproduction.
